mintscript/src/script/mod.rs

```diff
@@ -0,0 +1,228 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod verify;
+
+use common::chain::{signature::inputsig::InputWitness, timelock::OutputTimeLock, Destination};
+use utils::ensure;
+
+pub use verify::{ScriptError, ScriptErrorOf, ScriptResult, ScriptVisitor};
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum ScriptConstructionError {
+    #[error("Threshold requires {0} conditions but contains only {1}")]
+    InvalidThreshold(usize, usize),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Threshold {
+    required: usize,
+    conditions: Vec<ScriptCondition>,
+}
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq, Clone)]
+pub enum ThresholdError {
+    #[error("Not enough conditions satisfied")]
+    Insufficient,
+
+    #[error("Too many conditions declared satisfied")]
+    Excessive,
+}
+
+impl Threshold {
+    /// Trivially satisfied threshold condition (represented as 0-of-0 threshold)
+    pub const TRUE: Self = Self::new_unchecked(0, Vec::new());
+
+    /// Construct a threshold.
+    const fn new_unchecked(required: usize, conditions: Vec<ScriptCondition>) -> Self {
+        Self {
+            required,
+            conditions,
+        }
+    }
+
+    /// Construct a threshold.
+    ///
+    /// Creates an N-of-K threshold construct where N out of K conditions have to be satisfied.
+    /// Fails if N > K.
+    pub fn new(
+        required: usize,
+        conditions: Vec<ScriptCondition>,
+    ) -> Result<Self, ScriptConstructionError> {
+        ensure!(
+            required <= conditions.len(),
+            ScriptConstructionError::InvalidThreshold(required, conditions.len()),
+        );
+        Ok(Self::new_unchecked(required, conditions))
+    }
+
+    /// A [WitnessScript] consisting of only this threshold.
+    pub const fn into_script(self) -> WitnessScript {
+        WitnessScriptInner::Threshold(self).into_script()
+    }
+
+    /// Get all the conditions in this threshold construct.
+    pub fn conditions(&self) -> &[ScriptCondition] {
+        &self.conditions
+    }
+
+    /// Number of conditions that need to be satisfied in order for the threshold to be valid.
+    pub fn required(&self) -> usize {
+        self.required
+    }
+
+    /// Number of conditions that need to be dissatisfied in order for the threshold to be valid.
+    pub fn required_dissat(&self) -> usize {
+        let num_conds = self.conditions().len();
+        num_conds.checked_sub(self.required()).expect("checked during construction")
+    }
```

Comment:
> I mean, if no conditions are unsatisfied, the whole `Threshold` check will still succeed, right?

Though now I can see that `Threshold` requires the exact number of conditions to be satisfied. In that case `Threshold` itself is probably a misnomer.

Reply:
> Probably `required_dissat` is a misnomer, because it's the number of conditions that are allowed to fail. I mean, if no conditions are unsatisfied, the whole `Threshold` check will still succeed, right?

No, for malleability concerns, the user must pick the exactly N coditions, and satisfy them. Even if he can in principle satisfy more. This is in line with the usual multisig implementation in Taproot using the `OP_CHECKSIGADD` and `OP_EQUALSVERYFY` pattern.

> P.S. also, "unsatisfied" is IMO a better word here.

This is inspired by the terminology form miniscript and refers to the conscious choice of explicily marking a condition as not satisfied.